homeassistant/components/bluetooth/passive_update_processor.py

```diff
@@ -516,20 +518,39 @@ def remove_listener() -> None:
 
     @callback
     def async_update_listeners(
-        self, data: PassiveBluetoothDataUpdate[_T] | None
+        self,
+        data: PassiveBluetoothDataUpdate[_T] | None,
+        was_available: bool | None = None,
     ) -> None:
         """Update all registered listeners."""
+        if was_available is None:
+            was_available = self.coordinator.available
+
         # Dispatch to listeners without a filter key
         for update_callback in self._listeners:
             update_callback(data)
 
+        if not was_available or data is None:
+            # When data is None, or was_available is False,
+            # dispatch to all listeners as it means the device
+            # is flipping between available and unavailable
+            for listeners in self._entity_key_listeners.values():
+                for update_callback in listeners:
+                    update_callback(data)
+            return
+
         # Dispatch to listeners with a filter key
-        for listeners in self._entity_key_listeners.values():
-            for update_callback in listeners:
-                update_callback(data)
+        # if the key is in the data
+        entity_key_listeners = self._entity_key_listeners
+        for entity_key in data.entity_data:
+            if maybe_listener := entity_key_listeners.get(entity_key):
+                for update_callback in maybe_listener:
+                    update_callback(data)
```

Comment:
We knew what data had been updated but we were previously dispatching it to every entity 🤦